crates/storage/provider/src/providers/snapshot.rs

```diff
@@ -1,47 +1,211 @@
 use crate::{BlockHashReader, BlockNumReader, HeaderProvider, TransactionsProvider};
+use dashmap::DashMap;
 use reth_db::{
     table::{Decompress, Table},
     HeaderTD,
 };
 use reth_interfaces::{provider::ProviderError, RethResult};
-use reth_nippy_jar::{compression::Decompressor, NippyJar, NippyJarCursor};
+use reth_nippy_jar::{NippyJar, NippyJarCursor};
 use reth_primitives::{
-    snapshot::SegmentHeader, Address, BlockHash, BlockHashOrNumber, BlockNumber, ChainInfo, Header,
-    SealedHeader, TransactionMeta, TransactionSigned, TransactionSignedNoHash, TxHash, TxNumber,
+    snapshot::{SegmentHeader, SNAPSHOT_BLOCK_NUMBER_CHUNKS},
+    Address, BlockHash, BlockHashOrNumber, BlockNumber, ChainInfo, Header, SealedHeader,
+    SnapshotSegment, TransactionMeta, TransactionSigned, TransactionSignedNoHash, TxHash, TxNumber,
     B256, U256,
 };
-use std::ops::RangeBounds;
+use std::{ops::RangeBounds, path::PathBuf};
+
+/// Alias type for each specific `NippyJar`.
+type NippyJarRef<'a> =
+    dashmap::mapref::one::Ref<'a, (u64, SnapshotSegment), NippyJar<SegmentHeader>>;
 
 /// SnapshotProvider
-///
-///  WIP Rudimentary impl just for tests
-/// TODO: should be able to walk through snapshot files/block_ranges
-/// TODO: Arc over NippyJars and/or NippyJarCursors (LRU)
-#[derive(Debug)]
-pub struct SnapshotProvider<'a> {
-    /// NippyJar
-    pub jar: &'a NippyJar<SegmentHeader>,
+#[derive(Debug, Default)]
+pub struct SnapshotProvider {
+    /// Maintains a map which allows for concurrent access to different `NippyJars`, over different
+    /// segments and ranges.
+    map: DashMap<(BlockNumber, SnapshotSegment), NippyJar<SegmentHeader>>,
```

Comment:
nit: I think it will be more intuitive to have a block range as a key, because currently this key means a snapshot for `key..key+SNAPSHOT_BLOCK_NUMBER_CHUNKS`